src/sync/sync_permission.cpp

```diff
@@ -157,7 +157,85 @@ void Permissions::set_permission(std::shared_ptr<SyncUser> user,
 {
-    const auto realm_url = user->server_url() + permission.path;
+    auto props = AnyDict{
+        {"userId", permission.condition.user_id},
+        {"realmUrl", user->server_url() + permission.path},
+        {"mayRead", permission.access != AccessLevel::None},
+        {"mayWrite", permission.access == AccessLevel::Write || permission.access == AccessLevel::Admin},
+        {"mayManage", permission.access == AccessLevel::Admin},
+    };
+    if (permission.condition.type == Permission::Condition::Type::KeyValue) {
+        props.insert({"metadataKey", permission.condition.key_value.first});
+        props.insert({"metadataValue", permission.condition.key_value.second});
+    }
+    auto cb = [callback=std::move(callback)](Object*, std::exception_ptr exception) {
+        callback(exception);
+    };
+    perform_async_operation("PermissionChange", std::move(user), std::move(cb), std::move(props), make_config);
+}
+
+void Permissions::delete_permission(std::shared_ptr<SyncUser> user,
+                                    Permission permission,
+                                    PermissionChangeCallback callback,
+                                    const ConfigMaker& make_config)
+{
+    permission.access = AccessLevel::None;
+    set_permission(std::move(user), std::move(permission), std::move(callback), make_config);
+}
+
+void Permissions::make_offer(std::shared_ptr<SyncUser> user,
+                             PermissionOffer offer,
+                             PermissionOfferCallback callback,
+                             const ConfigMaker& make_config)
+{
+    auto props = AnyDict{
+        {"expiresAt", std::move(offer.expiration)},
+        {"userId", user->identity()},
+        {"realmUrl", user->server_url() + offer.path},
+        {"mayRead", offer.access != AccessLevel::None},
+        {"mayWrite", offer.access == AccessLevel::Write || offer.access == AccessLevel::Admin},
+        {"mayManage", offer.access == AccessLevel::Admin},
+    };
+    perform_async_operation("PermissionOffer",
+                            std::move(user),
+                            make_handler_extracting_property("token", std::move(callback)),
+                            std::move(props),
+                            make_config);
+}
+
+void Permissions::accept_offer(std::shared_ptr<SyncUser> user,
+                               const std::string& token,
+                               PermissionOfferCallback callback,
+                               const ConfigMaker& make_config)
+{
+    perform_async_operation("PermissionOfferResponse",
+                            std::move(user),
+                            make_handler_extracting_property("realmUrl", std::move(callback)),
+                            AnyDict{ {"token", token} },
+                            make_config);
+}
+
+Permissions::AsyncOperationHandler Permissions::make_handler_extracting_property(std::string property,
```

Comment:
Is there any reason for this to be a static member function rather than just a static function in the .cpp file?